File: module2/class1/exemplo7.cpp
/*
================================================================================
ATIVIDADE PRÁTICA 7 - PERFORMANCE DE ESCRITA EM DISCO (C++)
================================================================================

OBJETIVO:
- Demonstrar ineficiência de escritas byte-a-byte no disco
- Usar I/O profiling tools para identificar gargalos de disco
- Otimizar usando buffer de escrita
- Medir ganho de performance em operações de I/O

PROBLEMA:
- Escritas individuais de bytes causam múltiplas syscalls
- Cada write() é uma operação custosa do sistema operacional
- I/O Profiler mostrará alta latência e baixo throughput

SOLUÇÃO:
- Usar buffer para acumular dados e escrever em blocos maiores
- Resultado: redução drástica no número de syscalls

================================================================================
*/

#include <iostream>
#include <fstream>
#include <chrono>
#include <string>
using namespace std;

void inefficientDiskWrite(const string& filename, int dataSize) {
    cout << "Starting inefficient disk write..." << endl;
    cout << "Monitor I/O performance - should see many small write operations" << endl;
    
    auto start = chrono::high_resolution_clock::now();
    
    ofstream file(filename, ios::binary);
    if (!file.is_open()) {
        cerr << "Failed to open file!" << endl;
        return;
    }
    
    // PERFORMANCE ISSUE: Writing one byte at a time causes excessive syscalls
    for (int i = 0; i < dataSize; i++) {
        char byte = static_cast<char>(i % 256);
        file.write(&byte, 1); // Each write is a separate syscall - very inefficient!
        file.flush(); // Force immediate write to disk
        
        if (i % 10000 == 0) {
            cout << "Written " << i << "/" << dataSize << " bytes..." << endl;
        }
    }
    
    file.close();
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Inefficient write completed in: " << duration.count() << " ms" << endl;
    cout << "Total syscalls: ~" << dataSize << " (one per byte)" << endl << endl;
}

int main() {
    const int DATA_SIZE = 100000;
    const string filename = "test_output.dat";
    
    cout << "Starting disk I/O performance demonstration..." << endl;
    cout << "Task: Writing " << DATA_SIZE << " bytes to disk" << endl;
    cout << "Monitor I/O profiling tools for disk usage patterns" << endl << endl;
    
    inefficientDiskWrite(filename, DATA_SIZE);
    
    cout << "=== I/O PERFORMANCE ANALYSIS ===" << endl;
    cout << "Check I/O profiler for:" << endl;
    cout << "- High number of write syscalls" << endl;
    cout << "- Low I/O throughput" << endl;
    cout << "- High I/O wait time" << endl;
    
    return 0;
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO OTIMIZADA)
================================================================================

#include <iostream>
#include <fstream>
#include <chrono>
#include <string>
#include <vector>
using namespace std;

void efficientDiskWrite(const string& filename, int dataSize) {
    cout << "Starting efficient disk write..." << endl;
    cout << "Monitor I/O performance - should see fewer, larger write operations" << endl;
    
    auto start = chrono::high_resolution_clock::now();
    
    ofstream file(filename, ios::binary);
    if (!file.is_open()) {
        cerr << "Failed to open file!" << endl;
        return;
    }
    
    const int BUFFER_SIZE = 8192; // 8KB buffer
    vector<char> buffer;
    buffer.reserve(BUFFER_SIZE);
    
    // CORREÇÃO: Buffer writes to reduce syscalls
    for (int i = 0; i < dataSize; i++) {
        char byte = static_cast<char>(i % 256);
        buffer.push_back(byte);
        
        // Write buffer when full
        if (buffer.size() >= BUFFER_SIZE) {
            file.write(buffer.data(), buffer.size());
            buffer.clear();
        }
        
        if (i % 10000 == 0) {
            cout << "Buffered " << i << "/" << dataSize << " bytes..." << endl;
        }
    }
    
    // Write remaining data
    if (!buffer.empty()) {
        file.write(buffer.data(), buffer.size());
    }
    
    file.close();
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::milliseconds>(end - start);
    
    cout << "Efficient write completed in: " << duration.count() << " ms" << endl;
    cout << "Total syscalls: ~" << (dataSize / BUFFER_SIZE + 1) << " (buffered)" << endl;
}

int main() {
    const int DATA_SIZE = 100000;
    const string filename = "test_output.dat";
    
    cout << "Starting optimized disk I/O demonstration..." << endl;
    cout << "Task: Writing " << DATA_SIZE << " bytes to disk efficiently" << endl;
    cout << "Monitor I/O profiling tools for improved performance" << endl << endl;
    
    efficientDiskWrite(filename, DATA_SIZE);
    
    cout << "=== OPTIMIZED I/O RESULTS ===" << endl;
    cout << "Improvements:" << endl;
    cout << "- Dramatically fewer syscalls" << endl;
    cout << "- Higher I/O throughput" << endl;
    cout << "- Reduced I/O wait time" << endl;
    
    return 0;
}

================================================================================
*/

File: module2/class1/exemplo7.cs
/*
================================================================================
ATIVIDADE PRÁTICA 7 - PERFORMANCE DE ESCRITA EM DISCO (C#)
================================================================================

OBJETIVO:
- Demonstrar ineficiência de escritas byte-a-byte no disco
- Usar I/O profiling tools para identificar gargalos de disco
- Otimizar usando BufferedStream
- Medir ganho de performance em operações de I/O

PROBLEMA:
- Escritas individuais de bytes causam múltiplas syscalls
- FileStream.WriteByte() é uma operação custosa por byte
- I/O Profiler mostrará alta latência e baixo throughput

SOLUÇÃO:
- Usar BufferedStream para acumular dados e escrever em blocos
- Resultado: redução drástica no número de syscalls

================================================================================
*/

using System;
using System.IO;
using System.Diagnostics;

class Program {
    static void InefficientDiskWrite(string filename, int dataSize) {
        Console.WriteLine("Starting inefficient disk write...");
        Console.WriteLine("Monitor I/O performance - should see many small write operations");
        
        var sw = Stopwatch.StartNew();
        
        using (var file = new FileStream(filename, FileMode.Create, FileAccess.Write)) {
            // PERFORMANCE ISSUE: Writing one byte at a time causes excessive I/O operations
            for (int i = 0; i < dataSize; i++) {
                byte value = (byte)(i % 256);
                file.WriteByte(value); // Each write is potentially a separate I/O operation
                file.Flush(); // Force immediate write to disk
                
                if (i % 10000 == 0) {
                    Console.WriteLine($"Written {i}/{dataSize} bytes...");
                }
            }
        }
        
        sw.Stop();
        
        Console.WriteLine($"Inefficient write completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Approximate I/O operations: ~{dataSize} (one per byte)");
        Console.WriteLine();
    }
    
    static void Main() {
        const int DATA_SIZE = 100000;
        const string filename = "test_output.dat";
        
        Console.WriteLine("Starting disk I/O performance demonstration...");
        Console.WriteLine($"Task: Writing {DATA_SIZE} bytes to disk");
        Console.WriteLine("Monitor I/O profiling tools for disk usage patterns");
        Console.WriteLine();
        
        InefficientDiskWrite(filename, DATA_SIZE);
        
        Console.WriteLine("=== I/O PERFORMANCE ANALYSIS ===");
        Console.WriteLine("Check I/O profiler for:");
        Console.WriteLine("- High number of I/O operations");
        Console.WriteLine("- Low I/O throughput");
        Console.WriteLine("- High I/O wait time");
    }
}

/*
================================================================================
VERSÃO CORRIGIDA (DESCOMENTE PARA USAR A VERSÃO OTIMIZADA)
================================================================================

using System;
using System.IO;
using System.Diagnostics;

class Program {
    static void EfficientDiskWrite(string filename, int dataSize) {
        Console.WriteLine("Starting efficient disk write...");
        Console.WriteLine("Monitor I/O performance - should see fewer, larger write operations");
        
        var sw = Stopwatch.StartNew();
        
        using (var fileStream = new FileStream(filename, FileMode.Create, FileAccess.Write))
        using (var bufferedStream = new BufferedStream(fileStream, 8192)) { // CORREÇÃO: 8KB buffer
            for (int i = 0; i < dataSize; i++) {
                byte value = (byte)(i % 256);
                bufferedStream.WriteByte(value); // Buffered - reduces actual I/O operations
                
                if (i % 10000 == 0) {
                    Console.WriteLine($"Buffered {i}/{dataSize} bytes...");
                }
            }
            // BufferedStream automatically flushes on disposal
        }
        
        sw.Stop();
        
        Console.WriteLine($"Efficient write completed in: {sw.ElapsedMilliseconds} ms");
        Console.WriteLine($"Approximate I/O operations: ~{dataSize / 8192 + 1} (buffered)");
    }
    
    static void Main() {
        const int DATA_SIZE = 100000;
        const string filename = "test_output.dat";
        
        Console.WriteLine("Starting optimized disk I/O demonstration...");
        Console.WriteLine($"Task: Writing {DATA_SIZE} bytes to disk efficiently");
        Console.WriteLine("Monitor I/O profiling tools for improved performance");
        Console.WriteLine();
        
        EfficientDiskWrite(filename, DATA_SIZE);
        
        Console.WriteLine("=== OPTIMIZED I/O RESULTS ===");
        Console.WriteLine("Improvements:");
        Console.WriteLine("- Dramatically fewer I/O operations");
        Console.WriteLine("- Higher I/O throughput");
        Console.WriteLine("- Reduced I/O wait time");
    }
}

================================================================================
*/
